tests/test_update_query.py

```diff
@@ -97,7 +97,7 @@ def test_with_query_empty_str():
 
 def test_with_query_empty_value():
     url = URL("http://example.com/")
-    assert str(url.with_query({"a": ""})) == "http://example.com/?a="
+    assert str(url.with_query({"a": ""})) == "http://example.com/?a"
```

Comment:
What is the rationale for this change? I can't find anything in the URI or HTTP spec that suggests this normalisation.

Reply:
Rendering empty parameters with assignment does not work with the rfc3986 test cases:

https://github.com/aio-libs/yarl/blob/6b13867652ebcfad668f6e19146cce0b7ae4bc2c/tests/test_url.py#L1604

would be 
```
http://a/b/c/g?y=#s
```
which fails the unit tests

Reply:
I don't think it says anything about how to format the query string. The examples simply show the query unmodified. As far as RFC3986 is concerned, a query string is just a sequence of characters, and therefore shouldn't be modified during normalisation.

Reply:
Alternative I considered was using ?a= to store "" and ?a to store None and keep rendering the empty string value "" using an empty assignment. Adds quite some complexity for no additional value and using None would break the use case to remove a query parameter as discussed in https://github.com/aio-libs/yarl/issues/1010.

Reply:
> a query string is just a sequence of characters, and therefore shouldn't be modified during normalisation.

The unit tests of rfc3986 use parameters without assignment, serialization has to render those without assignment.
yarl does not have a notion for query parameters which been assigned with empty value or without assignment.
Due to the lack of notion if a parameters was assigned with assignment or without, the only option is to serialize all parameters without value as parameters without assignment.

Previously ?a was serialized as ?a=, now it is the other way around.

Reply:
> The unit tests of rfc3986 use parameters without assignment, serialization has to render those without assignment.

No, what I'm saying is that they show a query string (`y`) and after applying normalisation to that query string it still results in `y`. The RFC has no concept of parameters or assignments, it's just a string of characters that exclude `#`.

For encoding a set of parameters into a query string, we need to look elsewhere. I've found these steps from the W3C:
https://www.w3.org/TR/2014/REC-html5-20141028/forms.html#url-encoded-form-data

When presented with a form containing an empty 'y' parameter, these don't appear to mention or provide any steps that would result in a query string of `y`. It would always produce `y=`, so I think this is the best approach for serialising parameters, that should ensure best compatibility with other software.

Reply:
The unit tests of rfc3986 are join operations, with yarl is this includes encoding the parameters to have a string representation of the final url, this would add an `=` for empty parameters.

Proceed with adjusting the unit tests "the other direction"?

Reply:
In this particular example, there's no join, just a replacing of the query string from a dict of parameters. So, yes, I think move to producing `?a=`.

Reply:
This not about this particular example, but about encoding parameters in general.
Once it's a URL, the general encoding strategy is applied.

I adjusted the expectations of the rfc3986 unit tests to make it work accordingly.

Reply:
Hmm, I'm guessing the issue is that we are decoding the query string to parameters and then reencoding it. Ideally the rfc3986 tests should not need changing as the query string is a string literal in the URL and should therefore remain untouched. But, the .with_query() etc. code that encodes parameters should result in a `y=` format.

@webknjaz Any thoughts on how this is approached.

Reply:
Using query_string now, no encode required, query remains.